[MLIR] Adds a runtime switch to trigger MLIR

manupak opened the pull request:
Currently, in order to use MLIR, we need to rebuild the MIGraphX with a CMake flag.

This commit changes this have an switched polarity in the cmake flag -- i.e. MIGRAPHX_DISABLE_MLIR_BUILD. Then adds ENV var MIGRAPHX_MLIR to enable MLIR offloads in the fuse_mlir pass.

closes : https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/810

pfultz2 commented on src/targets/gpu/fuse_mlir.cpp:
```diff
@@ -35,7 +36,9 @@ struct module;
 namespace gpu {
 
-#ifdef MIGRAPHX_MLIR
+MIGRAPHX_DECLARE_ENV_VAR(MIGRAPHX_MLIR);
```
Should be called `MIGRAPHX_ENABLE_MLIR`.

manupak replied:
Ack

pfultz2 commented on Jenkinsfile:
```diff
@@ -118,7 +118,7 @@ rocmtest clang_debug: rocmnode('vega') { cmake_build ->
         def sanitizers = "undefined"
         def debug_flags = "-g -O2 -fsanitize=${sanitizers} -fno-sanitize-recover=${sanitizers}"
-        cmake_build(flags: "-DCMAKE_BUILD_TYPE=debug -DMIGRAPHX_ENABLE_PYTHON=Off -DMIGRAPHX_ENABLE_MLIR=On -DCMAKE_CXX_FLAGS_DEBUG='${debug_flags}' -DCMAKE_C_FLAGS_DEBUG='${debug_flags}'")
+        cmake_build(flags: "-DCMAKE_BUILD_TYPE=debug -DMIGRAPHX_ENABLE_PYTHON=Off -DCMAKE_CXX_FLAGS_DEBUG='${debug_flags}' -DCMAKE_C_FLAGS_DEBUG='${debug_flags}'")
```
MLIR still needs to be enabled at runtime.

pfultz2 commented:
This will also break packages being built since it wont build MLIR. MLIR needs to be added to requirements.txt(and removed from the dockerfile). However, we need to make sure we dont build MLIR on GH actions.

manupak commented:
> This will also break packages being built since it wont build MLIR. MLIR needs to be added to requirements.txt(and removed from the dockerfile). However, we need to make sure we dont build MLIR on GH actions. It would be better to just download prebuilt packages from repo.radeon.com.

I see. I think our packaging process is a bit WIP.

How about I re-instate the explicit enabled cmake flag and also keep the runtime switch ? (so that we can do measurements w/o needing to rebuild).

cc: @sjw36

pfultz2 commented:
> How about I re-instate the explicit enabled cmake flag and also keep the runtime switch ?

Sounds good to me.

manupak commented:
@pfultz2 , I ve made the changes now. PTAL when you have some time.

manupak commented:
Im trying to get this PR past CI...

https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/actions/runs/4373827517

Does anyone know what the following mean:

[release / MIGraphX Performance](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/actions/runs/4373827517/jobs/7654288286#step:5:28)
Input required and not supplied: token

pfultz2 commented on test/gpu/mlir.cpp:
```diff
@@ -187,4 +187,38 @@ module {
+    auto relu = m->add_instruction(migraphx::make_op("relu"), add);
+    m->add_return({relu});
+    EXPECT(verify_mlir_e2e(p));
+}
```
We already have verify tests for this.

manupak replied:
Sorry, I don't follow. Where would those test be?
I wanted to add full case which will also check the fuse_mlir pass

pfultz2 commented on test/gpu/mlir.cpp:
```diff
@@ -187,4 +187,38 @@ module {
+
+bool verify_mlir_e2e(const migraphx::program& p)
+{
+    setenv("MIGRAPHX_ENABLE_MLIR", "1", 1);
```
This should be set in the Jenkinsfile.

manupak commented:
(Here is an attempt to read between the lines)

> https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/actions/runs/4373827517
> 
> Does anyone know what the following mean:
> 
> [release / MIGraphX Performance](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/actions/runs/4373827517/jobs/7654288286#step:5:28) Input required and not supplied: token

This might potentially be coming from the fact the PR originates from my fork. I found that all other PRs in this project and pushed to this repo. So I had to create a new PR : https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/1610

> We already have verify tests for this.
> This should be set in the Jenkinsfile.

OK, I think I dont know/understand how the CI works here. So the comments kind of mean there is an CI (other GH Actions) that uses the jenkinsfile and runs the all tests 'MLIR Debug'. It would be great to know where to look for this CI.
On the assumption, I just removed e2e test I've added because I was thinking fuse_mlir is not tested.

Finally, this PR is supersceded by : https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/1610 on the assumption the CI checkout failure is due to the PR branch being in fork.

If my current assumptions are wrong, please do let me know.

manupak commented:
> OK, I think I dont know/understand how the CI works here. So the comments kind of mean there is an CI (other than GH Actions) that uses the jenkinsfile and runs the all tests 'MLIR Debug'. It would be great to know where to look for this CI.

ok got it -- somehow, I didn't see them being run in the PR. I can see the jenkins runs now.
